[addon-resizer] feat: Automatic reload nanny configuration

AhmedGrati opened the pull request:
#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind feature

#### What this PR does / why we need it:
Introduce automatic reload of nanny configuration.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #5700.

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
N/A
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->


- KEP: https://github.com/kubernetes/autoscaler/tree/master/addon-resizer/enhancements/5700-nanny-configuration-reload

mwielgus commented on addon-resizer/nanny/main/pod_nanny.go:
```diff
@@ -111,10 +113,55 @@ func main() {
+		Client:         k8s,
+	}
+
+	runNanny(*n, *nannyConfigurationFromFlags, *configDir)
```
nannyConfigurationFromFlags and configDir doesn't change, can we have them in the Nanny struct?

AhmedGrati replied:
Updated!

mwielgus commented on addon-resizer/nanny/main/pod_nanny.go:
```diff
@@ -111,10 +113,55 @@ func main() {
+	}
+}
+
+func runNanny(n nanny.Nanny, nannyCfdFlags nannyconfigalpha.NannyConfiguration, configDir string) {
```
Seems we could make it a nany function.

AhmedGrati replied:
Updated!

mwielgus commented on addon-resizer/nanny/main/pod_nanny.go:
```diff
@@ -111,10 +113,55 @@ func main() {
+	go n.PollAPIServer()
+}
+
+func getNannyEstimator(n nanny.Nanny, nannyCfdFlags nannyconfigalpha.NannyConfiguration, configDir string) (nanny.ResourceEstimator, error) {
```
Seems we could make it a Nanny function.

mwielgus commented on addon-resizer/nanny/utils/fswatcher.go:
```diff
@@ -0,0 +1,159 @@
+)
+
+// FsWatcher is a wrapper helper for watching files
+type FsWatcher struct {
```
Can we get some tests for this, please?

AhmedGrati replied:
Updated!

AhmedGrati commented:
@mwielgus Thanks for the review. I addressed your comments. For the `FsWatcher` unit tests, I added unit tests for `add` and `initWatcher` methods. For the `watch` method it needs root access to watch config files, which can be tested manually using this image `ahmedgrati/addon-resizer-amd64:1.8.19`. wdyt?

AhmedGrati commented:
ping @mwielgus. Can you please drop another quick review on this?
